src/backend/distributed/commands/function.c

```diff
@@ -91,6 +131,277 @@ CreateFunctionDDLCommandsIdempotent(const ObjectAddress *functionAddress)
+							errhint("Either provide a valid function argument name "
+									"or a valid \"$paramIndex\" to "
+									"create_distributed_function()")));
+		}
+
+		/*
+		 * Internal representation for the distributionArgumentIndex
+		 * starts from 0 whereas user facing API starts from 1.
+		 */
+		distributionArgumentIndex -= 1;
+	}
+
+	/*
+	 * If the user provided "$paramIndex", we're unnecessarily looping, but we prefer
+	 * this for the simplicity of the code.
+	 */
+	for (argIndex = 0; argIndex < numberOfArgs; ++argIndex)
+	{
+		char *argNameOnIndex = argNames != NULL ? argNames[argIndex] : NULL;
+		Oid argOidOnIndex = argTypes[argIndex];
+
+		/* if user specified as "$paramIndex" OR distribution_arg_name */
+		if (distributionArgumentIndex == argIndex ||
+			(argNameOnIndex != NULL &&
+			 pg_strncasecmp(argNameOnIndex, distributionArgumentName, NAMEDATALEN) == 0))
+		{
+			distributionArgumentIndex = argIndex;
+
+			*distributionArgumentOid = argOidOnIndex;
+
+			/* we found, no need to continue */
+			break;
+		}
+	}
+
+	/* we still couldn't find the argument, so error out */
+	if (distributionArgumentIndex == -1)
+	{
+		char *functionName = get_func_name(functionOid);
+
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("cannot distribute the function \"%s\" since the "
+							   "distribution argument is not valid ", functionName),
+						errhint("Either provide a valid function argument name "
+								"or a valid \"$paramIndex\" to "
+								"create_distributed_function()")));
+	}
+
+	ReleaseSysCache(proctup);
+
+	Assert(*distributionArgumentOid != InvalidOid);
+
+	return distributionArgumentIndex;
+}
+
+
+/*
+ * GetFunctionColocationId gets the parameters for deciding the colocationId
+ * of the function that is being distributed. The function errors out if it is
+ * not possible to assign a colocationId to the input function.
+ */
+static int
+GetFunctionColocationId(Oid functionOid, text *colocateWithText,
+						Oid distributionArgumentOid)
+{
+	int colocationId = INVALID_COLOCATION_ID;
+	char *colocateWithTableName = text_to_cstring(colocateWithText);
+	bool createdColocationGroup = false;
+
+	/*
+	 * Get an exclusive lock on the colocation system catalog. Therefore, we
+	 * can be sure that there will no modifications on the colocation table
+	 * until this transaction is committed.
+	 */
+	Relation pgDistColocation = heap_open(DistColocationRelationId(), ExclusiveLock);
+
+	if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) == 0)
+	{
+		/* check for default colocation group */
+		colocationId = ColocationId(ShardCount, ShardReplicationFactor,
+									distributionArgumentOid);
+
+		if (colocationId == INVALID_COLOCATION_ID)
+		{
+			colocationId =
+				CreateColocationGroup(ShardCount, ShardReplicationFactor,
+									  distributionArgumentOid);
+
+			createdColocationGroup = true;
+		}
+	}
+	else if (pg_strncasecmp(colocateWithTableName, "none", NAMEDATALEN) == 0)
```

Comment:
Well, I'm not sure if `none` makes sense for a function. But, we might keep it to have the same API with `create_distributed_table`

Reply:
I'd be in favour of removing it. More code more problems.